### PR TITLE
Mysql integration - work with the freeradius sql module

### PR DESCRIPTION
**Pull request by baimard**
Useful Mysql / Mariadb function to work with a database. That replaces the file yubikey_mappings. The mapping is made in the database. That's very useful when you work with freeradius which work with an SQL module.

I hope this will suit you.

Sorry for all the movement ...

### CODE REVIEW

**Comment by baimard** [comment]
I have fix some issues, but I can't get it to compile for Apple système (Mac OS). Don't know why ...

File: util.c [hunk 1]
```diff
@@ -52,6 +54,8 @@
 #include <ykdef.h>
 #endif /* HAVE_CR */
 
+#define STRING_SIZE 64
```
Comment (klali) [inline util.c:57]: this define isn't used?

File: util.c [hunk 2]
```diff
@@ -93,6 +97,192 @@ get_user_cfgfile_path(const char *common_path, const char *filename, const struc
...
+ * 
+ * CREATE TABLE IF NOT EXISTS  `otp`.`yubikey_mappings` ( 
+ *   `otp_id` VARCHAR(12) NOT NULL ,  
+ *   `username` VARCHAR(64) NOT NULL ,     
+ *   PRIMARY KEY  (`otp_id`(12))
+ *  );
+ *
+ */
+int
+check_user_token_mysql (const char *mysql_server,
```
Comment (klali) [inline util.c:115]: There's some trailing whitespace in this function, please get rid of those.

Some if statements don't have {} please add that.

File: util.c [hunk 3]
```diff
@@ -93,6 +97,192 @@ get_user_cfgfile_path(const char *common_path, const char *filename, const struc
...
+    }
+
+  const char *sql = "SELECT count(username) FROM yubikey_mappings WHERE username = ?;";
+  const char *sql2 = "SELECT count(username) FROM yubikey_mappings WHERE username = ? and otp_id = ?;";
+
+  if(otp_id == NULL)
+  {
+    if (mysql_stmt_prepare(stmt, sql, strlen(sql)))
+    {
+      fprintf(stderr, " mysql_stmt_prepare() failed\n");
```
Comment (klali) [inline util.c:185]: I'm not sure it makes sense to put these errors on stderr, the debug file is probable better. Try to keep them on one line rather than break it and put the mysql_stmt_error() on the next.

File: util.c [hunk 4]
```diff
@@ -93,6 +97,192 @@ get_user_cfgfile_path(const char *common_path, const char *filename, const struc
...
+      fprintf(stderr, " %s\n", mysql_stmt_error(stmt));
+      return retval; 
+    }
+  }
+
+
+  str_username= strlen(username);  
+  memset(ps_params, 0, sizeof(ps_params));
+  ps_params[0].buffer_type = MYSQL_TYPE_STRING;
+  ps_params[0].buffer = (char *)username;
```
Comment (klali) [inline util.c:202]: What's the guarantees here? would it be better to copy username to a local buffer and use that? Especially since we claim that the buffer is 64 bytes here.

File: util.c [hunk 5]
```diff
@@ -109,15 +299,15 @@ check_user_token (const char *authfile,
 		  const char *username,
 		  const char *otp_id,
 		  int verbose,
-                  FILE *debug_file)
+      FILE *debug_file)
```
Comment (klali) [inline util.c:302]: Please avoid non-related changes like these (even if they make sense). If you want to make them create a separate PR with only that.

File: util.h [hunk 6]
```diff
@@ -98,4 +99,4 @@ int challenge_response(YK_KEY *yk, int slot,
 size_t filter_result_len(const char *filter, const char *user, char *output);
 char *filter_printf(const char *filter, const char *user);
 
-#endif /* __PAM_YUBICO_UTIL_H_INCLUDED__ */
+#endif /* __PAM_YUBICO_UTIL_H_INCLUDED__ */
```
Comment (klali) [inline util.h:102]: Same here, unrelated change.

**Review by klali — CHANGES_REQUESTED** [review]
Overall I think this looks good, some minor comments.

File: util.h (hunk 6 above)
Comment (baimard) [reply util.h:102]: Alright, taken into account

File: util.c (hunk 5 above)
Comment (baimard) [reply util.c:302]: Alright, taken into account

File: util.c (hunk 3 above)
Comment (baimard) [reply util.c:185]: Sure ! Alright, taken into account

File: pam_yubico.c [hunk 7]
```diff
@@ -874,6 +886,16 @@ parse_cfg (int flags, int argc, const char **argv, struct cfg *cfg)
 	cfg->mode = CLIENT;
       if (strncmp (argv[i], "chalresp_path=", 14) == 0)
 	cfg->chalresp_path = argv[i] + 14;
+    //Mysql
```
[the remark this answers is not in the document]
Comment (baimard) [reply pam_yubico.c:889]: Alright, taken into account

File: util.c [hunk 8]
```diff
@@ -93,6 +97,192 @@ get_user_cfgfile_path(const char *common_path, const char *filename, const struc
...
+ *   PRIMARY KEY  (`otp_id`(12))
+ *  );
+ *
+ */
+int
+check_user_token_mysql (const char *mysql_server,
+    const char *mysql_user,
+    const char *mysql_password,
+    const char *mysql_database,
+		const char *username,
```
[the remark this answers is not in the document]
Comment (baimard) [reply util.c:119]: Alright, taken into account

File: pam_yubico.c [hunk 9]
```diff
@@ -134,6 +134,11 @@ struct cfg
   const char *user_attr;
   const char *yubi_attr;
   const char *yubi_attr_prefix;
+  //Mysql
```
[the remark this answers is not in the document]
Comment (baimard) [reply pam_yubico.c:137]: Alright, taken into account

File: configure.ac [hunk 10]
```diff
@@ -183,6 +183,12 @@ AC_LINK_IFELSE([AC_LANG_PROGRAM([[]],
 
 AC_CONFIG_FILES(Makefile)
 AC_CONFIG_FILES(tests/Makefile)
+
+#Mysql
+WITH_MYSQL()
```
[the remark this answers is not in the document]
Comment (baimard) [reply configure.ac:188]: done

File: pam_yubico.8.txt [hunk 11]
```diff
@@ -116,6 +116,18 @@ CA certitificate file for the LDAP connection.
 *chalresp_path*=_path_::
 Path of a system-wide directory where challenge-response files can be found for users. Default location is `$HOME/.yubico/`.
 
+*mysql_server*=_mysqlserver_::
+Uri of mysql server. Example 10.0.0.1
```
[the remark this answers is not in the document]
Comment (baimard) [reply pam_yubico.8.txt:120]: hostname/adress. you're right thank you !

File: util.c (hunk 2 above)
Comment (baimard) [reply util.c:115]: Also the "if(verbose)" ?

File: util.c (hunk 4 above)
Comment (baimard) [reply util.c:196]: I have change this party by :

```
  str_username = strlen(username);  
  memset(ps_params, 0, sizeof(ps_params));
  ps_params[0].buffer_type = MYSQL_TYPE_STRING;
  ps_params[0].buffer = (char *)username;
  ps_params[0].buffer_length = str_username;
  ps_params[0].length = &str_username;
  ps_params[0].is_null = 0;
```

The mysql documentation is not really clear about the usage of length et buffer_length.
If you really want a local buffer, I can do it. But I prefer like I have wrote.

File: pam_yubico.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (baimard) [reply pam_yubico.c]: I have make all the mysql code conditional like the ldap CODE.

File: util.c [hunk 12]
```diff
@@ -52,6 +52,10 @@
 #include <ykdef.h>
 #endif /* HAVE_CR */
 
+#if HAVE_MYSQL
```
Comment (klali) [inline util.c:55]: please use #ifdef here

File: util.c [hunk 13]
```diff
@@ -93,6 +97,196 @@ get_user_cfgfile_path(const char *common_path, const char *filename, const struc
   return 1;
 }
 
+#if HAVE_MYSQL
```
Comment (klali) [inline util.c:100]: #ifdef instead

File: util.c (hunk 2 above)
Comment (klali) [reply util.c:116]: not important, everything is already a mess.. =)

File: util.c (hunk 4 above)
Comment (klali) [reply util.c:196]: that const has to be thrown away on username might be a hint. I'm fine with this if it seems to be how others use this?

File: util.h [hunk 14]
```diff
@@ -51,6 +51,9 @@
 #define AUTH_NOT_FOUND -1 /* The requested token is not associated to the user */
 
 int get_user_cfgfile_path(const char *common_path, const char *filename, const struct passwd *user, char **fn);
+#if HAVE_MYSQL
```
Comment (klali) [inline util.h:54]: please switch to #ifdef and move to leftmost column

**Review by klali — CHANGES_REQUESTED** [review]

File: ax_compare_version.m4 [hunk 15]
```diff
@@ -0,0 +1,164 @@
+dnl (from http://autoconf-archive.cryp.to/ax_compare_version.m4 )
```
[the remark this answers is not in the document]
Comment (baimard) [reply ax_compare_version.m4:1]: done

File: util.c (hunk 12 above)
Comment (baimard) [reply util.c:55]: should I just replace #if with #ifdef?

Comment (klali) [reply util.c:55]: Yes, after dropping setting HAVE_MYSQL to 0 in the case where we don't have it

Comment (baimard) [reply util.c]: done

File: util.c (hunk 13 above)
Comment (baimard) [reply util.c]: done

File: configure.ac [hunk 16]
```diff
@@ -75,6 +75,20 @@ AC_ARG_WITH([ldap],
...
+              [disable support for mysql])],
+            [],
+            [with_mysql=yes])
+AS_IF([test "x$with_mysql" != xno],
+  [
+    PKG_CHECK_MODULES([MYSQL], [mysqlclient])
+    AC_DEFINE([HAVE_MYSQL], [1],[Define if you have mysqlclient])
+  ],
+  [
+    AC_DEFINE([HAVE_MYSQL], [0],[Define if you have mysqlclient])
```
[the remark this answers is not in the document]
Comment (baimard) [reply configure.ac]: tell me if it's ok, after next commit.

File: util.c [hunk 17]
```diff
@@ -137,8 +137,8 @@ check_user_token_mysql (const char *mysql_server,
   long unsigned int length;
   int int_data;
   int row_count;
-  bool is_null = false;
-  bool error = false;
+  bool is_null;
+  bool error;
```
Comment (baimard) [inline util.c]: Just fix these variables to follow the recommendation of mysql documentation

File: configure.ac (hunk 16 above)
[the remark this answers is not in the document]
Comment (klali) [reply configure.ac]: Almost, if you put AC_MSG_WARN() as a last argument to PKG_CHECK_MODULES() it will give that message and continue if libmysqlclient isn't found.

[the remark this answers is not in the document]
Comment (baimard) [reply configure.ac]: Ok done in the last commit.

File: util.c (hunk 2 above)
Comment (klali) [reply util.c:116]: there is still trailing whitespace.

**Review by klali — CHANGES_REQUESTED** [review]
I've mostly unresolved conversations where I don't believe we're 100% yet. The configuration part looks good.

File: configure.ac [hunk 18]
```diff
@@ -194,4 +208,4 @@ AC_MSG_NOTICE([Summary of build options:
   Library types:      Shared=${enable_shared}, Static=${enable_static}
   LDAP:               ${with_ldap}
   Challenge-Response: ${with_cr}
-])
+])
```
[the remark this answers is not in the document]
Comment (baimard) [reply configure.ac:210]: I'm not sure what modification you want here because for me this file have completely changed after last commit.

Could you please give me more precisions ?

File: configure.ac [hunk 19]
```diff
@@ -194,4 +207,4 @@ AC_MSG_NOTICE([Summary of build options:
...
```
Comment (klali) [inline configure.ac:210]: There an unnecessary (and wrong) change here removing the linefeed at the end of input, I don't want that in.

File: pam_yubico.c [hunk 20]
```diff
@@ -164,8 +169,19 @@ authorize_user_token (struct cfg *cfg,
 		      pam_handle_t *pamh)
 {
   int retval = AUTH_ERROR;
-
-  if (cfg->auth_file)
+  if (cfg->mysql_server)
+    {
+      #if HAVE_MYSQL
```
Comment (klali) [inline pam_yubico.c:174]: This should also be #ifdef

File: util.c [hunk 21]
```diff
@@ -93,6 +97,196 @@ get_user_cfgfile_path(const char *common_path, const char *filename, const struc
...
+ * 
+ * CREATE TABLE IF NOT EXISTS  `otp`.`yubikey_mappings` ( 
+ *   `otp_id` VARCHAR(12) NOT NULL ,  
+ *   `username` VARCHAR(64) NOT NULL ,     
+ *   PRIMARY KEY  (`otp_id`(12))
+ *  );
+ *
+ */
+int
+check_user_token_mysql (const char *mysql_server,
```
Comment (klali) [inline util.c:116]: This function needs to have indentation lineup in the header and not have trailing whitespace in the body.

File: util.c [hunk 22]
```diff
@@ -109,15 +303,15 @@ check_user_token (const char *authfile,
 		  const char *username,
 		  const char *otp_id,
 		  int verbose,
-                  FILE *debug_file)
+      FILE *debug_file)
```
Comment (klali) [inline util.c:306]: This is an unrelated change, please remove it from the PR

File: util.c [hunk 23]
```diff
@@ -109,15 +303,15 @@ check_user_token (const char *authfile,
...
 {
   char buf[1024];
   char *s_user, *s_token;
   int retval = AUTH_ERROR;
   int fd;
   struct stat st;
   FILE *opwfile;
-
+
```
Comment (klali) [inline util.c:314]: this is an unrelated change, please remove it from the PR:

File: util.c [hunk 24]
```diff
@@ -152,7 +346,7 @@ check_user_token (const char *authfile,
     {
       char *saveptr = NULL;
       if (buf[strlen (buf) - 1] == '\n')
-	buf[strlen (buf) - 1] = '\0';
+	      buf[strlen (buf) - 1] = '\0';
```
Comment (klali) [inline util.c:349]: this is an unrelated change, please remove it from the PR:

File: util.c [hunk 25]
```diff
@@ -598,4 +792,4 @@ char *filter_printf(const char *filter, const char *user) {
   char *result = malloc(filter_result_len(filter, user, NULL));
   filter_result_len(filter, user, result);
   return result;
-}
+}
```
Comment (klali) [inline util.c:795]: this is an unrelated change, please remove it from the PR:

File: util.h [hunk 26]
```diff
@@ -98,4 +101,4 @@ int challenge_response(YK_KEY *yk, int slot,
 size_t filter_result_len(const char *filter, const char *user, char *output);
 char *filter_printf(const char *filter, const char *user);
 
-#endif /* __PAM_YUBICO_UTIL_H_INCLUDED__ */
+#endif /* __PAM_YUBICO_UTIL_H_INCLUDED__ */
```
Comment (klali) [inline util.h:104]: this is an unrelated change, please remove it from the PR.

**Review by klali — CHANGES_REQUESTED** [review]
I resolved all conversations and opened new comments in all places where I think there's a change needed.

**Comment by baimard** [comment]
> I resolved all conversations and opened new comments in all places where I think there's a change needed.

thanks, I appreciate it. I understand that the code have to been clean !

**Review by baimard — COMMENTED** [review]
I think all are resolved. :)

File: pam_yubico.c [hunk 27]
```diff
@@ -164,8 +169,19 @@ authorize_user_token (struct cfg *cfg,
 		      pam_handle_t *pamh)
 {
   int retval = AUTH_ERROR;
-
-  if (cfg->auth_file)
+  if (cfg->mysql_server)
+    {
+      #ifdef HAVE_MYSQL
```
Comment (klali) [inline pam_yubico.c:174]: please move this to the leftmost column

**Review by klali — CHANGES_REQUESTED** [review]

File: pam_yubico.c (hunk 27 above)
Comment (baimard) [reply pam_yubico.c]: Sorry for the delay, I haven't see your comment.

done in the last commit.

**Comment by klali** [comment]
Great. I'm ok to merge this. One final thing. I want you to squash everything to one commit so we skip alot of the comment/refactor noise on master.

I'd recommend doing this with a "git rebase -i master" and marking all commits except the first as squash then finally amending that commit with a good commit message.

**Comment by baimard** [comment]
> Great. I'm ok to merge this. One final thing. I want you to squash everything to one commit so we skip alot of the comment/refactor noise on master.
> 
> I'd recommend doing this with a "git rebase -i master" and marking all commits except the first as squash then finally amending that commit with a good commit message.
> 
> edit: and after squashing it you'll have to force-push the branch.

Thank you, I have squash all the commits except the first and I have change the commit message. Let me see if that well for you.

I am very happy with this collaboration !

Thank you again for all reviewing and all your constructive comments.

**Comment by klali** [comment]
Thanks!

I've merged this, it's notable that the CI should be moved to run on github actions and include tests for this, but lets leave that as an exercise for the future.
